### ISSUE TITLE
[RFC] FSDP arguments refactoring

### ISSUE THREAD

**Issue by min-xu-ai** [facebookresearch/fairscale#684]
cc @myleott @sshleifer @QuentinDuval @prigoyal @anj-s @blefaudeux @msbaines @tmarkstrum @SeanNaren 

# Motivations

Currently FSDP has a long list of arguments. They deal with flattening, sharding, mixed precision, cpu offloading. The long list is harder for users to get a quick glance on what's supported combinations and what's not. It is harder for developers to reasonable about interactions and unit testing all combinations.

# Proposal

I propose we separate the params into different groups, as the following:

```
parameter_handling: enum of "flatten", "original"
sharding: enum of "none, full, shard_after_backward"
mixed_precision: enum of "full, amp_w32_b32_g32, amp_w16_b16_g16, ..."
cpu_offloading: enum of "none, grad, param, grad_and_param"
```

The combination space above is still very big but at least the related arguments are grouped and it is easier to see what's supported and what's not.

# Implementation

We can first put the new arguments in place and deprecate the older ones. Some of the options may raise NotImplemented error and we will add implementation support gradually.

I am learning from Benjamin where we can use an enum class that's also a string, like:

```
class Foo(str, Enum):
   V1 = "val1"
```

This way, we can have both yaml config and enum based argument passing.

Any comments or suggestions are most welcome from the cc'ed folks and the community. Please add people if I missed anyone.

**Comment by blefaudeux** [comment]
another option, fairly typical in C or C++ (but not really in python/pytorch) is to pass a single config structure for which you can provide some defaults and catch incompatibilities at construction time (@dataclass is pretty powerful for these things). Not saying that it's universally better, can be an annoyance also for people, adds another layer, just throwing the idea around

**Comment by blefaudeux** [comment]
1 more cent on the enum/str (on top of the serialization/deserialization, built in assert and single place to see what's supported vs. manual testing of strings in some places): makes it cleaner unit test wise, since you can @parameterize on the enum values, and any new case you add in the future will be tested automatically.

**Comment by min-xu-ai** [comment]
Thanks @blefaudeux! I guess you are in favor of this change? :-)

I thought about @DataClass as well. It seems like an overkill at the moment. It definitely helps with configuration. But it may make object construction a bit more indirect since users have to create a config class first. I think at a higher configuration level (i.e. beyond a single FSDP instance, configurations of the entire model with nested FSDP instances, perhaps we will need config like dataclass. But I certainly haven't really thought through that). For this proposal, I was thinking to keep it simple and concise with a few enum arguments.

**Comment by blefaudeux** [comment]
> Thanks @blefaudeux! I guess you are in favor of this change? :-)
> 
> I thought about @DataClass as well. It seems like an overkill at the moment. It definitely helps with configuration. But it may make object construction a bit more indirect since users have to create a config class first. I think at a higher configuration level (i.e. beyond a single FSDP instance, configurations of the entire model with nested FSDP instances, perhaps we will need config like dataclass. But I certainly haven't really thought through that). For this proposal, I was thinking to keep it simple and concise with a few enum arguments.

certainly fine by me, it was just so that multiple options were considered, the config does feel a bit heavy handed indeed

**Comment by SeanNaren** [comment]
Thanks @min-xu-ai great RFC :)

Just playing devils advocate (as I've run into this myself), how stable do you feel the parameters are? Do you expect them to change much? Over the release we saw quite a few additional parameters as you guys were figuring out convergence issues etc. Just making sure we don't over complicate if we plan for these variables to disappear in the future.

I've definitely lost touch of the FSDP configurations (I've seen many PRs be merged on edge cases for FP16 and whatnot, but I have no clue what should be enabled and when) so I do support this RFC to try readable defaults :)

**Comment by min-xu-ai** [comment]
Thanks for the comments and support of the RFC, @SeanNaren!

I feel this will be the stable parameters moving forward. With the enums, we will be adding supported values and potentially deprecating the unsupported ones. For example:

mixed_precision: "full", "amp1", "amp2", ...

We might be adding more values to the enum, like adding support of "amp3" mode or removing support of "amp1" mode, but the parameter "mixed_precision" will stay the same. This way, I think the API is stable at least.

Yeah, part of the refactor is to make user's life easier figure out what should enabled and when for what they want to do. String values help, so is the separation of the arguments into different topic, like params, sharding, mixed precision and offloading. There are still a few more options beyond these 4 topics. So it is still complicated to config, unfortunately. Good news is that we have iron out most issues with big vision models and FSDP is running well and fast for them now! :-)

**Comment by myleott** [comment]
I like the motivation and this proposal generally makes sense to me.

One concern is that this may complicate the internal code if we have to check multiple values. For example, instead of:
```python
if self.mixed_precision:
   (...)
```
would we now have something like this?
```python
if self.mixed_precision in {"amp_w32_b32_g32", "amp_w16_b16_g16", ...}:
   (...)
```
The latter seems messier to me, but perhaps you could just remap these strings to the more intricate settings in the `__init__`?

**Comment by min-xu-ai** [comment]
> The latter seems messier to me, but perhaps you could just remap these strings to the more intricate settings in the __init__?

Yes, internally it will still be various flags. The API change is only to make it easier to config and to reason what are the possible cases we will have to support and test. For example, with 2 bool arguments, we will have 4 cases max. But it is some times the case that 1 of the combinations doesn't make sense and/or isn't useful (like mixed_precision=False and compute_dtype=fp16), so we can shorten the number of supported cases. I.e. 3 supported string values on a single argument.

A potential downside of this is that there could be too many string values. We end up with a big string -> tuple conversion. My intent is try to limit the number of supported cases. It definitely requires discipline to not to add too many cases just because we can. :-)

**Comment by myleott** [comment]
Sounds good

> But it is some times the case that 1 of the combinations doesn't make sense and/or isn't useful (like mixed_precision=False and compute_dtype=fp16)

Btw, we do actually use that combination in fairseq 😄 That corresponds to pure FP16 training, which is exposed in fairseq as `--memory-efficient-fp16`

**Comment by min-xu-ai** [comment]
> Btw, we do actually use that combination in fairseq 😄 That corresponds to pure FP16 training, which is exposed in fairseq as `--memory-efficient-fp16`

Do you not run that under autocast? Are the weights first promoted to fp32 and then matmul with fp32 inputs?

**Comment by myleott** [comment]
> Do you not run that under autocast? Are the weights first promoted to fp32 and then matmul with fp32 inputs?

No autocast,. We manually cast some ops to FP32 (e.g., any softmax in the network), but otherwise it's pure FP16. The key to making this converge for Transformer is to do the optimizer step in FP32, although we actually store the Adam stats in a scaled FP16 format for memory efficiency. This is similar to what OpenAI did with Jukebox.

**Comment by min-xu-ai** [comment]
> No autocast. We cast some ops manually to FP32 (e.g., any softmax in the network), but otherwise it's pure FP16.

This is pretty amazing. A simple matmul can "easily" overflow if the sum of the element-wise multiply from a row vector and a column vector goes over 65536. Do you guys use dynamic loss scaling and/or drop samples in training if it causes any inf/nan in forward or backward in that mode though?

**Comment by min-xu-ai** [comment]
To make the discussion more concrete, here is what the new options could look like. Please let me know what do you think.

- Does this improve user's experiences?
- Does it obfuscate the simple boolean or dtype options?
- Will this be more likely a more stable API for things like lightning?

I am totally OK to abandon if this turns out not as good as people think. :-)

https://github.com/facebookresearch/fairscale/blob/06cf7c78c686ec85019d220e215d9301f76d853e/fairscale/nn/data_parallel/fully_sharded_data_parallel.py#L44-L164

**Comment by min-xu-ai** [comment]
^ Any comments?

**Comment by anj-s** [comment]
Thank you @min-xu-ai for proposing this! This is definitely a step in usability and will be much appreciated by users. Do you have an example of how the API will change? I may have missed that snippet from above.

**Comment by prigoyal** [comment]
This is quite nice! Sharing one high-level thought that I think @myleott also mentioned earlier: `I16_W16_C16_B16_G32 ` patterns like these offer nice compactness but IMHO, makes it a bit "messy" + "magic strings" even though the strings are pretty nice if one knows the context. I think we should allow setting things via strings like this + additionally allow user to do settings in __init__ currently directly (could be nice for a new user). I am not too sure what the right path is but I think we can experiment and determine user preference. For VISSL, we might prefer the explicit settings in `__init__` purely because of readability.  :)

**Comment by min-xu-ai** [comment]
> Do you have an example of how the API will change?

You didn't miss it. I haven't put it in. It could be something like:

```
def __init__(self,
    shard: ShardingControl,
    param: ParameterControl,
    precision: PrecisionControl,
    offload: OffloadControl,
    # other misc options
    ...
    # other deprecated options
    ...
):
    ...
```

**Comment by min-xu-ai** [comment]
> For VISSL, we might prefer the explicit settings in __init__ purely because of readability.

Thanks @prigoyal. See the comment to @anj-s, the new options will be for `__init__` to replace too many bool or string options. For the long strings like `I16_W16_C16_B16_G32`, the hope is that out of 32 cases, we use comments to mark out a few commonly used ones so that users will use them primarily. We can even give them alias names. This way, it is less likely to make mistakes compared with setting 5 booleans in some particular value (1 copy & paste or type of a meaningful alias vs. 5 copy & paste or typing True/False...)

**Comment by anj-s** [comment]
> > Do you have an example of how the API will change?
> 
> You didn't miss it. I haven't put it in. It could be something like:
> 
> ```
> def __init__(self,
>     shard: ShardingControl,
>     param: ParameterControl,
>     precision: PrecisionControl,
>     offload: OffloadControl,
>     # other misc options
>     ...
>     # other deprecated options
>     ...
> ):
>     ...
> ```

Ah ok! I think the user example could be:
model = FSDP(ParamterControl.Flatten, PrecisionControl. I32_W32_C32_B32_G32, ....) 
OR
model = FSDP(param="flatten", precision="i32_w32_c32_b32_g32", ....)

Is that right?

Couple of questions:
- will the user have reasonable defaults for these args?
- would it make sense to break apart the PrecisionControl into two args. Seems overloaded given the number of combinations.
- Will we make ParameterControlFlatten and other enum structures public?

There are both pros and cons of this approach. The only thing for me is going forward if we have a new argument to FSDP should we create a new structure? If we don't and we end up with multiple args, then do we refactor into another Enum. Maybe we are confident that the API is stable now but future extensibility might be difficult with this approach.

**Comment by min-xu-ai** [comment]
> Is that right?

Exactly!

>     will the user have reasonable defaults for these args?

Yes, the defaults will be more informatively marked with the string values, like full precision is good for this, full fp16 is good for that, etc.

>    would it make sense to break apart the PrecisionControl into two args. Seems overloaded given the number of combinations.

The idea is to not make it common for users to use all 32 combinations. We may end up eventually just support the most useful handful of them. My intent is to help with usability and lower the burden on the user so that they don't need to know implementation details. So, the idea is keep this a single argument that's easy to use (i.e. the common options could be just "full_precision", "full_fp16", "mixed_network_optimized", "mixed_fast_network", etc. Those are alias into some specific I32..., I16 options.)

>    Will we make ParameterControlFlatten and other enum structures public?

They don't have to since they are subclass of str. The only use case for them to be publish is for unit test to enumerate them like `@parameterized(option, list(ParameterControl))`, which make the unit test easily enumerate all supported options.

The idea is that these 4 are stable options. New things will most likely be a new enum for one of the 4 options. Therefore, I don't foresee adding more new options beyond these 4 after they are introduced.

**Comment by min-xu-ai** [comment]
Closing this since Yanli has taken some spirit from this RFC and API change while upstreaming FSDP to pytorch is a better route. I think it is OK to let FSDP API evolve in fairscale.
